### PR TITLE
feat(user-tenant-service): adds user tenant service

### LINKED ISSUE #623: @sourceloop/user-tenant-service - To add this new service
**Is your feature request related to a problem? Please describe.**
User-tenant-service is most common service, which is used in approx all projects, it can be part of sourceloop-services, using same current auth-db.

**Describe the solution you'd like**
Basic user-tenant-service (logical-separation) can be part of sourceloop service for ready-to-use APIs

These should be Default APIs

**User-Tenant Service APIs** 

**Roles APIs (Only super-admin will have access to these APIs)**

- [ ] GET /roles
- [ ] POST /roles
- [ ] GET /roles/count
- [ ] PUT /roles/{id}
- [ ] PATCH /roles/{id}
- [ ] GET /roles/{id}
- [ ] DELETE /roles/{id}
- [ ] PATCH /roles

**Tenant APIs (Only super-admin can do CRUD of tenants)**

- [ ] GET /tenants/count
- [ ] GET /tenants/{id}/config
- [ ] GET /tenants/{id}/users/count
- [ ] PATCH /tenants/{id}/users/{userId}
- [ ] DELETE /tenants/{id}/users/{userId}
- [ ] GET /tenants/{id}/users/{userid}
- [ ] POST /tenants/{id}/users
- [ ] GET /tenants/{id}/users
- [ ] PATCH /tenants/{id}
- [ ] GET /tenants/{id}
- [ ] DELETE /tenants/{id}
- [ ] POST /tenants
- [ ] PATCH /tenants
- [ ] GET /tenants
- [ ] GET /user-tenants/{id}

**Tenant-Roles APIs**	

- [ ] GET /roles/{id}/user-tenants
- [ ] POST /roles/{id}/user-tenants
- [ ] PATCH /roles/{id}/user-tenants
- [ ] DELETE /roles/{id}/user-tenants
- [ ] GET /roles/{id}/user-tenants/count
- [ ] POST /user-tenant-prefs
- [ ] GET /user-tenant-prefs
- [ ] GET /user-tenants/{id}

**Groups APIs
(These tables are already in db-migrations, so we can keep these APIs)**

- [ ] GET /groups/count
- [ ] GET /groups/{id}/user-groups/count
- [ ] PATCH /groups/{id}/user-groups/{userGroupId}
- [ ] DELETE /groups/{id}/user-groups/{userGroupId}
- [ ] POST /groups/{id}/user-groups
- [ ] GET /groups/{id}/user-groups
- [ ] PATCH /groups/{id}
- [ ] GET /groups/{id}
- [ ] DELETE /groups/{id}
- [ ] POST /groups
- [ ] GET /groups
- [ ] GET /user-groups/count
- [ ] GET /user-groups

### PR DESCRIPTION
**Pull request by sumiter92**
## Description

Adds user tenant service

Fixes  #623 

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] Performed a self-review of my own code
- [ ] Code conforms with the style guide

### CODE REVIEW

File: services/user-tenant-service/src/controllers/ping.controller.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  @authorize({permissions: ['*']})
+  @get('/health', {
+    responses: {
+      [STATUS_CODE.OK]: PING_RESPONSE,
+    },
+  })
+  async health(): Promise<object> {
+    // try {
```
Comment (akshatdubeysf) [inline services/user-tenant-service/src/controllers/ping.controller.ts:59]: ??

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

File: services/user-tenant-service/README.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
+# @sourceloop/payment-service
```
[the remark this answers is not in the document]
Comment (sumiter92) [reply services/user-tenant-service/README.md]: done

File: services/user-tenant-service/src/controllers/ping.controller.ts (hunk 1 above)
Comment (sumiter92) [reply services/user-tenant-service/src/controllers/ping.controller.ts]: removed

File: services/user-tenant-service/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Check out https://hub.docker.com/_/node to select a new base image
+FROM node:10-slim
```
Comment (akshatdubeysf) [inline services/user-tenant-service/Dockerfile:2]: Dockerfile not required in a component

File: services/user-tenant-service/package.json [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    "url": "https://github.com/sourcefuse/loopback4-microservice-catalog.git",
+    "directory": "services/user-tenant-service"
+  },
+  "author": "Sourcefuse",
+  "license": "MIT",
+  "files": [
+    "README.md",
+    "dist",
+    "src",
+    "migration.js",
```
Comment (akshatdubeysf) [inline services/user-tenant-service/package.json:51]: This file is missing?

File: services/user-tenant-service/public/index.html [hunk 5]
```diff
@@ -0,0 +1,88 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+  <title>services</title>
```
Comment (akshatdubeysf) [inline services/user-tenant-service/public/index.html:5]: user-tenant-service

Comment (sumiter92) [reply services/user-tenant-service/public/index.html]: done

File: services/user-tenant-service/package.json (hunk 4 above)
Comment (sumiter92) [reply services/user-tenant-service/package.json]: done

File: services/user-tenant-service/src/controllers/ping.controller.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
+import {inject} from '@loopback/core';
+// import {get, HttpErrors, ResponseObject} from '@loopback/rest';
```
Comment (akshatdubeysf) [inline services/user-tenant-service/src/controllers/ping.controller.ts:2]: ????

File: services/user-tenant-service/src/controllers/ping.controller.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import {get, ResponseObject} from '@loopback/rest';
+import {CONTENT_TYPE, ILogger, LOGGER, STATUS_CODE} from '@sourceloop/core';
+import {authorize} from 'loopback4-authorization';
+// import {PgdbDataSource} from '../datasources';
```
Comment (akshatdubeysf) [inline services/user-tenant-service/src/controllers/ping.controller.ts:6]: ???

File: services/user-tenant-service/src/controllers/ping.controller.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    },
+  },
+};
+
+/**
+ * A simple controller to bounce back http requests
+ */
+export class PingController {
+  constructor(
+    // @inject('datasources.pgdb') private readonly dataSource: PgdbDataSource,
```
Comment (akshatdubeysf) [inline services/user-tenant-service/src/controllers/ping.controller.ts:32]: ??

File: services/user-tenant-service/src/controllers/role-user-tenant.controller.ts [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+  requestBody,
+} from '@loopback/rest';
+import {
+  CONTENT_TYPE,
+  STATUS_CODE,
+  OPERATION_SECURITY_SPEC,
+} from '@sourceloop/core';
+import {authenticate, STRATEGY} from 'loopback4-authentication';
+import {authorize} from 'loopback4-authorization';
+// import {PermissionKey} from '@local/rakuten-core';
```
Comment (akshatdubeysf) [inline services/user-tenant-service/src/controllers/role-user-tenant.controller.ts:26]: ??

File: services/user-tenant-service/src/controllers/user-group.controller.ts [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        limit: 1,
+      });
+      if (userGroupRecord && !userGroup?.isOwner) {
+        throw new HttpErrors.Forbidden('The team needs at least one owner');
+      }
+    }
+    await this.userGroupService.updateById(userGroupId, userGroup);
+
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    this.groupRepository.updateById(id, {
```
Comment (akshatdubeysf) [inline services/user-tenant-service/src/controllers/user-group.controller.ts:189]: Why is this added manually?

Comment (akshatdubeysf) [reply services/user-tenant-service/src/controllers/user-group.controller.ts:189]: Use usermodifiablecrud repo

File: services/user-tenant-service/src/models/group.model.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
+// import {UserTenantGroupType} from '@local/rakuten-core';
```
Comment (akshatdubeysf) [inline services/user-tenant-service/src/models/group.model.ts:1]: ??

**Review by akshatdubeysf — APPROVED** [review]

**Review by akshatdubeysf — APPROVED** [review]

File: services/user-tenant-service/src/controllers/user-signup.controller.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
+import {PermissionKey, AuthProvider} from '../enums';
+import {repository} from '@loopback/repository';
+import {get, getModelSchemaRef, param} from '@loopback/rest';
+import {CONTENT_TYPE, ErrorCodes, STATUS_CODE} from '@sourceloop/core';
+import {authenticate, STRATEGY} from 'loopback4-authentication';
+import {authorize} from 'loopback4-authorization';
+import {UserSignupCheckDto} from '../models';
+import {UserCredentialsRepository, UserRepository} from '../repositories';
+export const PMS_SIGN_UP_RESPONSE_DESCRIPTION = 'Success Response.';
```
[the remark this answers is not in the document]
Comment (sumiter92) [reply services/user-tenant-service/src/controllers/user-signup.controller.ts]: done , changed

**Pull request merged**